File: 2023/1/part_two.py
from pathlib import Path

from part_one import get_calibration_value


def parse_document(document: str) -> int:
    return sum(get_calibration_value(convert_line(line)) for line in document.strip().split('\n'))

def convert_line(line: str) -> str:
  d = {}
  for idx, _ in enumerate(line):
    if line[idx:idx + 3] == 'one':
      d[idx] = '1'
    elif line[idx:idx + 3] == 'two':
      d[idx] = '2'
    elif line[idx:idx + 5] == 'three':
      d[idx] = '3'
    elif line[idx:idx + 4] == 'four':
      d[idx] = '4'
    elif line[idx:idx + 4] == 'five':
      d[idx] = '5'
    elif line[idx:idx + 3] == 'six':
      d[idx] = '6'
    elif line[idx:idx + 5] == 'seven':
      d[idx] = '7'
    elif line[idx:idx + 5] == 'eight':
      d[idx] = '8'
    elif line[idx:idx + 4] == 'nine':
      d[idx] = '9'
  for n, (pos, digit) in enumerate(d.items()):
    line = line[:pos + n] + digit + line[pos + n:]
  return line

if __name__ == '__main__':
  print(parse_document(Path('sample_input_2.txt').read_text()))
  print(parse_document(Path('document.txt').read_text()))

File: 2023/1/part_two.clj
(require '[clojure.string :as str])

(defn get-number [string]
  (let [first-number (re-find #"[\d+]" string)
        last-number (re-find #"[\d+](?!.*\d)" string)]
    (Integer/parseInt (str first-number last-number))))

(defn transform-line [line]
  (as-> line l
    (str/replace l #"one" "o1ne")
    (str/replace l #"two" "t2wo")
    (str/replace l #"three" "t3hree")
    (str/replace l #"four" "f4our")
    (str/replace l #"five" "f5ive")
    (str/replace l #"six" "s6ix")
    (str/replace l #"seven" "s7even")
    (str/replace l #"eight" "e8ight")
    (str/replace l #"nine" "n9ine")))

(defn main [file-name]
  (as-> (slurp file-name) x
      (str/split x #"\n")
      (map transform-line x)
      (map get-number x)
      (reduce + x)))

(main "document.txt")
